browser/net/global_privacy_control_network_delegate_helper_browsertest.cc

```diff
@@ -0,0 +1,145 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <map>
+
+#include "brave/common/brave_features.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "components/network_session_configurator/common/network_switches.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/browser_test_utils.h"
+#include "content/public/test/test_utils.h"
+#include "net/dns/mock_host_resolver.h"
+#include "net/test/embedded_test_server/http_request.h"
+#include "net/test/embedded_test_server/http_response.h"
+
+enum GPCHeaderResult {
+  OK,
+  NO_REQUEST,
+  NO_HEADER,
+  WRONG_VALUE,
+};
+
+class GlobalPrivacyControlNetworkDelegateBrowserTest
+    : public InProcessBrowserTest {
+ public:
+  GlobalPrivacyControlNetworkDelegateBrowserTest()
+      : https_server_(net::EmbeddedTestServer::TYPE_HTTPS) {
+    feature_list_.InitAndEnableFeature(features::kGlobalPrivacyControl);
+  }
+
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+
+    header_result_ = GPCHeaderResult::NO_REQUEST;
+
+    host_resolver()->AddRule("*", "127.0.0.1");
+
+    LOG(ERROR) << "Set up on main thread";
+
+    https_server_.RegisterRequestHandler(base::BindRepeating(
+        &GlobalPrivacyControlNetworkDelegateBrowserTest::HandleRequest,
+        base::Unretained(this)));
+
+    ASSERT_TRUE(https_server_.Start());
+  }
+
+  std::unique_ptr<net::test_server::HttpResponse> HandleRequest(
+      const net::test_server::HttpRequest& request) {
+    auto http_response =
```

Comment:
I don't think you really need to craft responses, the same can be achieved with `RegisterRequestMonitor` - e.g. see `BraveContentSettingsAgentImplBrowserTest`. Also this callback is called on IO thread, so acess to `header_result_` should be guarded